### PR TITLE
Refactoring with MicroBlock Block Hash

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

This PR is to enable the usage of BlockHash for MicroBlock. As a side effect, many of the logics for handling microblocks will also be simplified, includes:
- Check For MicroBlocks during FinalBlock Consensus
- LoadUnavailableMicroBlock when received FinalBlock from DS
- Lookup ProcessForwardTransaction
- Send MicroBlock To Lookup/Archival

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [x] small-scale cloud test

### CODE REVIEW

File: src/libData/BlockData/Block/MicroBlock.cpp [hunk 1]
```diff
@@ -81,6 +81,15 @@ MicroBlock::MicroBlock(const MicroBlockHeader& header,
   m_cosigs = move(cosigs);
 }
 
+MicroBlock::MicroBlock(const MicroBlockHeader& header,
+                       const vector<TxnHash>& tranHashes)
+    : m_header(header), m_tranHashes(tranHashes) {
+  if (m_header.GetNumTxs() != m_tranHashes.size()) {
```
Comment (ansnunez) [inline src/libData/BlockData/Block/MicroBlock.cpp]: We can probably have a more useful message here, like displaying `GetNumTxs` and `size` values.

File: src/libData/BlockData/Block/TxBlock.cpp [hunk 2]
```diff
@@ -72,14 +72,13 @@ TxBlock::TxBlock(const vector<unsigned char>& src, unsigned int offset) {
...
-                 const vector<MicroBlockHashSet>& microBlockHashes,
+                 const vector<BlockHash>& microBlockHashes,
                  const vector<uint32_t>& shardIds, CoSignatures&& cosigs)
     : m_header(header),
       m_isMicroBlockEmpty(isMicroBlockEmpty),
       m_microBlockHashes(microBlockHashes),
       m_shardIds(shardIds) {
-  if (m_header.GetNumMicroBlockHashes() != m_microBlockHashes.size() &&
-      m_header.GetNumMicroBlockHashes() != m_shardIds.size()) {
+  if (m_header.GetNumMicroBlockHashes() != m_microBlockHashes.size()) {
```
Comment (ansnunez) [inline src/libData/BlockData/Block/TxBlock.cpp]: Same comment here about the log message

File: src/libData/BlockData/Block/TxBlock.cpp [hunk 3]
```diff
@@ -89,14 +88,13 @@ TxBlock::TxBlock(const TxBlockHeader& header,
...
-                 const vector<MicroBlockHashSet>& microBlockHashes,
+                 const vector<BlockHash>& microBlockHashes,
                  const vector<uint32_t>& shardIds)
     : m_header(header),
       m_isMicroBlockEmpty(isMicroBlockEmpty),
       m_microBlockHashes(microBlockHashes),
       m_shardIds(shardIds) {
-  if (m_header.GetNumMicroBlockHashes() != m_microBlockHashes.size() &&
-      m_header.GetNumMicroBlockHashes() != m_shardIds.size()) {
+  if (m_header.GetNumMicroBlockHashes() != m_microBlockHashes.size()) {
```
Comment (ansnunez) [inline src/libData/BlockData/Block/TxBlock.cpp]: Same comment here about the log message

File: src/libMessage/ZilliqaMessage.proto [hunk 4]
```diff
@@ -137,27 +137,18 @@ message ProtoMicroBlock
...
 
 message ProtoTxBlock
 {
     message TxBlockHashSet
     {
-        required bytes txroothash               = 1;
+        required bytes mbroothash               = 1;
         required bytes stateroothash            = 2;
-        required bytes deltaroothash            = 3;
         required bytes statedeltahash           = 4;
```
Comment (ansnunez) [inline src/libMessage/ZilliqaMessage.proto]: Number has to be changed from 4 to 3

**Comment by ansnunez** [comment]
File src/libMessage/ZilliqaMessage.proto.full doesn't need to be committed

File: src/libDirectoryService/FinalBlockPreProcessing.cpp [hunk 5]
```diff
@@ -933,7 +884,7 @@ bool DirectoryService::CheckBlockHash() {
                 "DS committee hash in newly received Tx Block doesn't match. "
                 "Calculated: "
                     << committeeHash << " Received: "
-                    << m_pendingDSBlock->GetHeader().GetCommitteeHash());
+                    << m_finalBlock->GetHeader().GetCommitteeHash());
```
Comment (KaustubhShamshery) [inline src/libDirectoryService/FinalBlockPreProcessing.cpp]: This does not seem related to this PR? Was this a bug?
